lib/handlers/authorize-handler.ts

```diff
@@ -0,0 +1,377 @@
+      state = this.getState(request);
+      RequestedResponseType = this.getResponseType(request, client);
+      responseType = new RequestedResponseType(this.options);
+      const codeOrAccessToken = await responseType.handle(
+        request,
+        client,
+        user,
+        uri,
+        scope,
+      );
+      const redirectUri = this.buildSuccessRedirectUri(uri, responseType);
+      this.updateResponse(response, redirectUri, responseType, state);
+
+      return codeOrAccessToken;
+    } catch (e) {
+      if (!(e instanceof OAuthError)) {
+        e = new ServerError(e);
+      }
+
+      const redirectUri = this.buildErrorRedirectUri(uri, responseType, e);
+
+      this.updateResponse(response, redirectUri, responseType, state);
+
+      throw e;
+    }
+  }
+
+  /**
+   * Get the client from the model.
+   */
+
+  async getClient(request: Request) {
+    const clientId = request.body.client_id || request.query.client_id;
+
+    if (!clientId) {
+      throw new InvalidRequestError('Missing parameter: `client_id`');
+    }
+
+    if (!is.vschar(clientId)) {
+      throw new InvalidRequestError('Invalid parameter: `client_id`');
+    }
+
+    const redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+    if (redirectUri && !is.uri(redirectUri)) {
+      throw new InvalidRequestError(
+        'Invalid request: `redirect_uri` is not a valid URI',
+      );
+    }
+
+    const client = await this.model.getClient(clientId);
+    if (!client) {
+      throw new InvalidClientError(
+        'Invalid client: client credentials are invalid',
+      );
+    }
+
+    if (!client.grants) {
+      throw new InvalidClientError('Invalid client: missing client `grants`');
+    }
+
+    const responseType =
+      request.body.response_type || request.query.response_type;
+    const requestedGrantType =
+      responseType === 'token' ? 'implicit' : 'authorization_code';
+
+    if (!client.grants.includes(requestedGrantType)) {
+      throw new UnauthorizedClientError(
+        'Unauthorized client: `grant_type` is invalid',
+      );
+    }
+
+    if (!client.redirectUris || client.redirectUris.length === 0) {
+      throw new InvalidClientError(
+        'Invalid client: missing client `redirectUri`',
+      );
+    }
+
+    if (redirectUri && !client.redirectUris.includes(redirectUri)) {
+      throw new InvalidClientError(
+        'Invalid client: `redirect_uri` does not match client value',
+      );
+    }
+
+    return client;
+  }
+
+  /**
+   * Validate requested scope.
+   */
+  async validateScope(user: User, client: Client, scope: string) {
+    if (this.model.validateScope) {
+      const validatedScope = await this.model.validateScope(
+        user,
+        client,
+        scope,
+      );
+      if (!validatedScope) {
+        throw new InvalidScopeError(
+          'Invalid scope: Requested scope is invalid',
+        );
+      }
+
+      return validatedScope;
+    }
+
+    return scope;
```

Comment:
I think the code intends it to be so, but a model could just skip any validation with this code?

Reply:
if this.model.validateScope is not set, then we dont/cant validate the scope. 
So... what is your issue?